velox/exec/HashProbe.h

```diff
@@ -382,11 +382,16 @@ class HashProbe : public Operator {
     template <typename TOnMiss>
     void advance(vector_size_t row, bool passed, TOnMiss onMiss) {
       if (currentRow != row) {
-        if (currentRow != -1 && !currentRowPassed) {
+        if (currentRow != -1 && !currentRowPassed && !lastRowAdded) {
+          // Add the currentRow,i.e, the previous row for the incoming row
+          // with NULLs to build side. If the currentRow is from the previous
```

Comment:
| row | filterPassed |
| 0 | 1
| 1 | 0 <- `batch ended. your logic adds <1, NULL> to the output.`
| 1 | 1 <- `new batch. The previous entry into rawOutputMapping was wrong and we must have <1, output> in the final result`
| 1 | 0

Yes. So if we get the last row in a batch as with <row=1,filterPassed=0>, we just add this as <1, NULL> and update rawOutputProbeRowMapping as well. We then updated `noMatchDetector_.lastRow=true` for tracking. Also, in this batch the value of numPassed will be equal to numRows, which was not the case earlier. For example, it would have been 9 rather than 10 due to this last row being not added.
```
outputTableRows_[N] = nullptr;
rawOutputProbeRowMapping[N] = 1;
numPassed = N+1;
```

Now when we have the new batch first row with <row=1,filterPassed=1>, then we check if the condition `noMatchDetector_.lastRow=true` for the previous row, which by the way meets here. So we avoid doing anything to the `noMatchDetector_.currentRow` and update the new row values to them as well the `noMatchDetector_.lastRow=false`. Then, the control hits back to the passed criteria and does the update for this first row in the new batch.
```
if (1) {
        outputTableRows_[N+1] = outputTableRows_[0];
        rawOutputProbeRowMapping[N+1] = rawOutputProbeRowMapping[0];
        numPassed = N+1+1;
      }
```

So this doesn't handle the situation you cited above. The updated to rawOutputMapping will not be changed.

Reply:
Would you please help me understand how `<1, NULL>`  is a valid row in the final output? Can you write a test case where `<1, NULL>` and `1, <value>` both appear in the output for a left join with filter?